verification.py

```diff
@@ -4399,137 +4569,101 @@ def _run_pyvolve_comparison(self, model, length=20, num_samples=10):
+    def register(model):
+        group = "pyvolve"
+        name = f"{group}_{model}"
+        runner.register(PyvolveTest1(name=name, group=group, model=model))
 
+    register("JC69")
+    register("HKY")
 
-def get_test_names():
-    tests = []
 
-    def get_class_name(c):
-        return f"{c}"[17:-6].lower()
+@attr.s
+class TestRunner:
+    """
+    Class responsible for registering all known tests and running
+    them.
+    """
+
+    tests = attr.ib(init=False, default=[])
+    groups = attr.ib(init=False, default=set())
+
+    def register(self, test):
+        self.tests.append(test)
+        self.groups.add(test.group)
 
-    # use a simple DFS to get the names of classes and
-    classes = SimulationVerifier.__subclasses__()
-    while classes:
-        top = classes.pop()
-        tests.append(get_class_name(top))
-        for c in top.__subclasses__():
-            classes.append(c)
-    return tests
+    def run(self, output_dir, names=None, group=None):
+        if names != []:
+            tests = [test for test in self.tests if test.name in names]
+        elif group is not None:
+            tests = [test for test in self.tests if test.group == group]
+        else:
+            tests = self.tests
+        for test in tests:
+            test.run(output_dir)
 
 
-def add_simulator_arguments(parser):
+def add_simulator_arguments(parser, groups):
     parser.add_argument(
         "--extended",
         action="store_true",
         help="Run extended tests in dtwf_vs_coalescent",
     )
-    tests = get_test_names()
     parser.add_argument(
         "--group",
         "-g",
         default=None,
-        choices=tests,
+        choices=groups,
         help="Run all tests for specified group",
     )
     parser.add_argument("tests", nargs="*", help="Run specific tests")
 
 
-def run_tests(args):
+def run_tests(args, runner):
     output_dir = "tmp__NOBACKUP__"
-    classes = []
-    ms = MsTest(output_dir)
-    discoal = DiscoalTest(output_dir)
-    dtwf = DtwfVsCoalescentTest(output_dir)
-    xi = XiTest(output_dir)
-    hudson = HudsonTest(output_dir)
-    contdiscrete = ContDiscreteTest(output_dir)
-    argrecord = ArgRecordTest(output_dir)
-    random = RandomTest(output_dir)
-    sim = SimTest(output_dir)
-    analytical = AnalyticalTest(output_dir)
-    smc = SmcTest(output_dir)
-    seqgen = SeqGenTest(output_dir)
-    pyvolve = PyvolveTest(output_dir)
-
-    if args.group is None:
-        classes.extend(
-            [
-                dtwf,
-                xi,
-                hudson,
-                contdiscrete,
-                argrecord,
-                random,
-                sim,
-                analytical,
-                smc,
-                discoal,
-                ms,
-                seqgen,
-                pyvolve,
-            ]
-        )
+    if args.extended:
+        register_dtwfvscoalescent_tests(runner, add_random_dtwf=True)
+    if args.tests is not []:
```

Comment:
Comparing for identity isn't recommended. Better to use ``if len(args.tests) > 0``. Also this has broken the ``-g`` behaviour, which doesn't currently work.